docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,212 @@
+# [RFC] Conjure Endpoint Errors
+
+**Date:** 2021/03/19
+
+**Stakeholders:**
+* *Anyone that produces or consumes conjure-generated code.*
+
+* * *
+
+## **Context**
+
+Currently, errors can be defined in Conjure type definitions. These errors can then be thrown by services, and if a client 
+knows about them, they can be handled during runtime by clients. The exact handling of this mechanism may vary by
+the implementation (for example, in Java these are `ServicException`s, which become `RemoteException`s which can be caught
+by the client).
+
+There are multiple issues with this. As a consumer, you only know that an endpoint may throw an error by reading the 
+documentation or the implementation itself. As a producer, you are limited in the expressivity of what you can say 
+with your exception, as you have no guarantee that a consumer will handle your exception (or even read your documentation).
+
+This RFC introduces *declared errors* on Conjure endpoints. These errors represent real outcomes of a call to an endpoint, 
+and should be handled by consumers.
+
+## Existing Documentation
+
+### PRs for example implementation (in Java, using checked exceptions)
+
+* Root PR (Conjure API changes): https://github.com/palantir/conjure/pull/816
+* Conjure Java API Changes: https://github.com/palantir/conjure-java-runtime-api/pull/634
+* Conjure Java & Dialogue implementation: https://github.com/palantir/conjure-java/pull/1275
+
+## Example Conjure implementations
+
+**Conjure API**
+
+```
+types:
+  definitions:
+    default-package: com.palantir.my.api
+    errors:
+      BlueMoon:
+        namespace: MyConjureService
+        code: CUSTOM_SERVER
+
+services:
+  MyConjureService:
+    name: An example service.
+    package: com.palantir.my.api
+    default-auth: header
+    endpoints:
+      getSomeResult:
+        http: GET /
+        returns: list<rid>
+        errors:
+          - BlueMoon
+        docs: Returns some result, but once in a blue moon throws an error
+```
+## Proposal
+
+This proposal should be viewed in two distinct parts - declared (and handled) exceptions on Conjure endpoints, and the 
+language implementations of such a feature.
+
+### Intent
+
+The primary intent of this proposal is to provide more expressivity around errors in a Conjure API, while minimizing the 
+change surface-area, especially around wire format changes. Our proposed change introduces *no* wire changes, so should 
+be purely additive and not dramatically affect back-compat.
+
+### Declared Errors
```

Comment:
want this RFC to contain information about how this fits in with backwards compatibility guarantees

Reply:
@j-baker This RFC should not involve any backwards breaking changes - old clients will continue to throw exceptions in exactly the same way they do currently (as the wire format will not change at all).

I'll add a section on compat to emphasise this.

Reply:
It has always been the case that you cannot break APIs semantically and this is still  true  with the proposed change, however it may encourage people to "fix" their APIs by adding declared exceptions after the fact. This exact problem was alluded to in previous discussions but to highlight it here:

Suppose you had a `TryAgainLaterException` that you threw when you were being throttled. As a client you could have handled this nicely and just having this exception converted to declared doesn't have to be a big deal for client authors. However, now that there is a shiny new feature, a dev decides to add a `ThrottledException` that includes how long you should wait. This change would have always been a semantic break but now we have gone from a world where exceptions were by and large ignored to one where we are saying "Look Here. Declare your exceptions! It is backwards and forwards safe!" when the reality isn't quite so.

Reply:
@ericanderson As you point out, changing the type of an error is a semantic break, and always has been. Your fear is that adding a feature will make developers believe it's safe to make this kind of semantic break, that could be the case, however there are a couple ways of looking at it:
1. The proposal would make error usages more obvious in API review, thus reviewed more closely to avoid semantic breaks in ways that have always been easy to miss in code review.
2. The existing error type may be declared (improving endpoint documentation in a way that may allow better codegen, but no wire or semantic break)

In terms of backwards compatibility:
* Renaming an existing error is a break. This prevents existing clients from effectively handling a failure case that previously worked as expected.
* Removing an error _may_ not be a break. If an error type cannot be thrown anymore (inputs that were not allowed at one point are now allowed) it's reasonable to remove the error. Old clients which handled the failure won't execute error handling code because the server no longer produces that error. However if the same scenario results in a different kind of error, this is a break as described by the first bullet.
* Adding a new error is not a break. This assumes the error was not previously considered some other kind of error that clients may have been expected to handle. When the server may fail in a new way, clients may be asked to handle it. There's an expectation that the new failure cannot _require_ special handling from clients because old clients must continue to work, however this can be leveraged to provide better user-facing messages in the UI or to describe a failure that was previously surfaced as a 500 response that didn't provide sufficient context.

Note that removing an error may not be a break, but if another error is added in the future for an equivalent case but using a different name, we've created the "Renaming an existing error is a break." case.